### PR TITLE
Implementation of SizeData

### PR DESCRIPTION
**Pull request by hsyyid**
Implementation of `SizeData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableSizeData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `VALUE_PROCESSOR`(s):
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `FIELD_NAME` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

``` java
package io.github.hsyyid.testplugin;

import org.spongepowered.api.Sponge;
import org.spongepowered.api.block.BlockTypes;
import org.spongepowered.api.command.CommandException;
import org.spongepowered.api.command.CommandResult;
import org.spongepowered.api.command.CommandSource;
import org.spongepowered.api.command.args.CommandContext;
import org.spongepowered.api.command.args.GenericArguments;
import org.spongepowered.api.command.spec.CommandExecutor;
import org.spongepowered.api.command.spec.CommandSpec;
import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.entity.Entity;
import org.spongepowered.api.entity.EntityType;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.cause.Cause;
import org.spongepowered.api.event.game.state.GameInitializationEvent;
import org.spongepowered.api.plugin.Plugin;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.text.format.TextColors;
import org.spongepowered.api.util.blockray.BlockRay;
import org.spongepowered.api.util.blockray.BlockRayHit;
import org.spongepowered.api.world.Location;
import org.spongepowered.api.world.World;
import org.spongepowered.api.world.extent.Extent;

import java.util.Optional;

@Plugin(id = "TestPlugin", name = "TestPlugin", version = "1.0")
public class TestPlugin {

    @Listener
    public void onGameInit(GameInitializationEvent event) {
        CommandSpec sizeCommandSpec = CommandSpec.builder()
                .description(Text.of("Size Command"))
                .arguments(GenericArguments.seq(
                        GenericArguments.onlyOne(GenericArguments.string(Text.of("entity"))),
                        GenericArguments.onlyOne(GenericArguments.doubleNum(Text.of("height"))),
                        GenericArguments.onlyOne(GenericArguments.doubleNum(Text.of("width"))),
                        GenericArguments.onlyOne(GenericArguments.doubleNum(Text.of("scale")))))
                .executor(new TestExecutor())
                .build();

        Sponge.getGame().getCommandManager().register(this, sizeCommandSpec, "size");
    }

    public class TestExecutor implements CommandExecutor {

        public CommandResult execute(CommandSource src, CommandContext ctx) throws CommandException {
            String entity = ctx.<String>getOne("entity").get();
            double height = ctx.<Double>getOne("height").get();
            double width = ctx.<Double>getOne("width").get();
            double scale = ctx.<Double>getOne("scale").get();

            EntityType entityType = Sponge.getRegistry().getType(EntityType.class, entity).orElse(null);

            if (entityType != null) {
                if (src instanceof Player) {
                    Player player = (Player) src;
                    spawnEntity(getSpawnLocFromPlayerLoc(player).add(0, 1, 0), entityType, 1, height, width, scale);
                    player.sendMessage(Text.of(TextColors.GREEN, "Success! ", TextColors.YELLOW, "Spawned mob(s)!"));
                } else {
                    src.sendMessage(Text.of(TextColors.DARK_RED, "Error! ", TextColors.RED, "You're not an in-game player."));
                }
            } else {
                src.sendMessage(Text.of(TextColors.DARK_RED, "Error! ", TextColors.RED, "The entity doesn't exist."));
            }

            return CommandResult.success();
        }

        public void spawnEntity(Location<World> location, EntityType type, int amount, double height, double width, double scale) {
            for (int i = 1; i <= amount; i++) {
                Extent extent = location.getExtent();
                Optional<Entity> optional = extent.createEntity(type, location.getPosition());
                Entity entity = optional.get();
                entity.offer(Keys.HEIGHT, (float) height);
                entity.offer(Keys.BASE_SIZE, (float) width);
                entity.offer(Keys.SCALE, (float) scale);
                extent.spawnEntity(entity, Cause.of(this));
            }
        }

        public Location<World> getSpawnLocFromPlayerLoc(Player player) {
            BlockRay<World> playerBlockRay = BlockRay.from(player).blockLimit(350).build();

            BlockRayHit<World> finalHitRay = null;

            while (playerBlockRay.hasNext()) {
                BlockRayHit<World> currentHitRay = playerBlockRay.next();

                if (!player.getWorld().getBlockType(currentHitRay.getBlockPosition()).equals(BlockTypes.AIR)) {
                    finalHitRay = currentHitRay;
                    break;
                }
            }

            Location<World> spawnLocation = null;

            if (finalHitRay == null) {
                spawnLocation = player.getLocation();
            } else {
                spawnLocation = finalHitRay.getLocation();
            }

            return spawnLocation;
        }
    }
}
```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    protected SizeData createManipulator() {
+        return new SpongeSizeData();
+    }
+
+    @Override
+    protected boolean doesDataExist(Entity entity) {
+        return !entity.isDead;
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java]: The data still exists if the entity is dead.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/HealthValueProcessor.java [hunk 2]
```diff
@@ -63,7 +64,7 @@ protected boolean set(EntityLivingBase container, Double value) {
 
     @Override
     protected ImmutableValue<Double> constructImmutableValue(Double value) {
-        return new ImmutableSpongeBoundedValue<>(Keys.HEALTH, value, 20D, doubleComparator(), 0D, 20D);
+        return new ImmutableSpongeValue<>(Keys.HEALTH, value);
```
Comment (JBYoshi) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/HealthValueProcessor.java]: What's this change for?

File: src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    protected boolean set(Entity entity, Map<Key<?>, Object> keyValues) {
+        float scale = (Float) keyValues.get(Keys.SCALE);
+        entity.height = ((Float) keyValues.get(Keys.HEIGHT)) * scale;
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java]: I think you need to call `setSize` instead, which does things like updating the bounding box.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java (hunk 1 above)
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java]: I'm also not sure that this should be supported for all entities. It only makes sense on certain entites, and in those cases, it's usually only the scale which can be modified client-side.

Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java]: I agree, this should not be supported for all entities.

File: src/main/java/org/spongepowered/common/data/DataRegistrar.java [hunk 4]
```diff
@@ -155,6 +155,10 @@ public static void setupSerialization(Game game) {
         dataManager.registerDataProcessorAndImpl(IgniteableData.class, SpongeIgniteableData.class, ImmutableIgniteableData.class,
                 ImmutableSpongeIgniteableData.class, igniteableProcessor);
 
+        final SizeDataProcessor sizeProcessor = new SizeDataProcessor();
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/DataRegistrar.java]: Unneeded variable

**Comment by hsyyid** [comment]
@simon816 I'll get back on this in a bit

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeSizeData.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                .maximum(Float.MAX_VALUE)
+                .build()
+                .asImmutable();
+
+        registerGetters();
+    }
+
+    @Override
+    public ImmutableValue<Float> base() {
+        return new ImmutableSpongeValue<>(Keys.BASE_SIZE, this.base);
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeSizeData.java]: Store `ImmutableValue`s in fields

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeSizeData.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    @Override
+    public int compareTo(ImmutableSizeData o) {
+        return ComparisonChain.start()
+                .compare(o.base().get().intValue(), this.base)
+                .compare(o.height().get().intValue(), this.height)
+                .compare(o.scale().get().intValue(), this.scale)
+                .result();
+    }
+
+    public float getBase() {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeSizeData.java]: Methods not really needed as you can inline it with a lambda below
`registerFieldGetter(Keys.BASE_SIZE, () -> this.base);` etc

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeSizeData.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        registerFieldGetter(Keys.HEIGHT, SpongeSizeData.this::getHeight);
+        registerFieldSetter(Keys.HEIGHT, SpongeSizeData.this::setHeight);
+        registerKeyValue(Keys.HEIGHT, SpongeSizeData.this::height);
+
+        registerFieldGetter(Keys.SCALE, SpongeSizeData.this::getScale);
+        registerFieldSetter(Keys.SCALE, SpongeSizeData.this::setScale);
+        registerKeyValue(Keys.SCALE, SpongeSizeData.this::scale);
+    }
+    
+    public float getBase() {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeSizeData.java]: Methods can be inlined with lambdas

File: src/main/java/org/spongepowered/common/data/processor/value/entity/BaseSizeValueProcessor.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        return Optional.empty();
+    }
+
+    @Override
+    public boolean supports(ValueContainer<?> container) {
+        return container instanceof Entity;
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Float value) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/BaseSizeValueProcessor.java]: Why override `offerToStore` when `set(Entity, Float)` should be sufficient?

File: src/main/java/org/spongepowered/common/data/processor/value/entity/BaseSizeValueProcessor.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        return Optional.of(container.width);
+    }
+
+    @Override
+    protected ImmutableBoundedValue<Float> constructImmutableValue(Float value) {
+        return constructValue(value).asImmutable();
+    }
+
+    @Override
+    public Optional<MutableBoundedValue<Float>> getApiValueFromContainer(ValueContainer<?> container) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/BaseSizeValueProcessor.java]: `getVal(Entity)` should be sufficient instead

File: src/main/java/org/spongepowered/common/data/processor/value/entity/HeightValueProcessor.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        return Optional.of(container.height);
+    }
+
+    @Override
+    protected ImmutableBoundedValue<Float> constructImmutableValue(Float value) {
+        return constructValue(value).asImmutable();
+    }
+
+    @Override
+    public Optional<MutableBoundedValue<Float>> getApiValueFromContainer(ValueContainer<?> container) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/HeightValueProcessor.java]: `getVal(Entity)` should be sufficient

File: src/main/java/org/spongepowered/common/data/processor/value/entity/HeightValueProcessor.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                    .maximum(Float.MAX_VALUE)
+                    .defaultValue(1f)
+                    .actualValue(height)
+                    .build());
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public boolean supports(ValueContainer<?> container) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/HeightValueProcessor.java]: No need to override as passing `Entity.class` in the constructor should be all that's needed

File: src/main/java/org/spongepowered/common/data/processor/value/entity/HeightValueProcessor.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        return container instanceof Entity;
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Float value) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/HeightValueProcessor.java]: `set(Entity, Float)` should be all that's needed

File: src/main/java/org/spongepowered/common/data/processor/value/entity/ScaleValueProcessor.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        return Optional.of(1.0F);
+    }
+
+    @Override
+    protected ImmutableBoundedValue<Float> constructImmutableValue(Float value) {
+        return constructValue(value).asImmutable();
+    }
+
+    @Override
+    public Optional<MutableBoundedValue<Float>> getApiValueFromContainer(ValueContainer<?> container) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/ScaleValueProcessor.java]: No need as `getVal(Entity)` should be all that's needed

File: src/main/java/org/spongepowered/common/data/processor/value/entity/ScaleValueProcessor.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+                    .maximum(Float.MAX_VALUE)
+                    .defaultValue(1f)
+                    .actualValue(1.0F)
+                    .build());
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public boolean supports(ValueContainer<?> container) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/ScaleValueProcessor.java]: Passing `Entity.class` in the constructor already covers this

File: src/main/java/org/spongepowered/common/data/processor/value/entity/ScaleValueProcessor.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        return container instanceof Entity;
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Float value) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/ScaleValueProcessor.java]: `set(Entity, Float)` is all that's needed

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 16]
```diff
@@ -1143,6 +1143,18 @@ public void setUntargetable(boolean untargetable) {
         this.untargetable = untargetable;
     }
 
+    @Override
+    public void setSpongeSize(float width, float height, float scale) {
+        Entity spongeEntity = this;
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: Pointless variable

**Comment by simon816** [comment]
I see you're returning a constant `1` for scale, you may want to see how to get the scale from when I first implemented `getScale`
https://github.com/SpongePowered/SpongeForge/pull/141/files#diff-7935aaf1dc2aaf2359fbb6ef6905acf3 (My code has since been deleted it seems)

File: src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.data.entity;
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/SizeDataProcessor.java]: Should be in the `processor.multi.entity` package

**Comment by gabizou** [comment]
@hsyyid Can you update this pr by rebasing on bleeding?

**Comment by hsyyid** [comment]
@gabizou ok

**Comment by parlough** [comment]
@hsyyid Do you plan to continue this or should someone else take it up?

**Comment by hsyyid** [comment]
@Meronat Someone else can take this

**Comment by kashike** [comment]
Superseded by #1492.

**Pull request closed without merging**